clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java

```diff
@@ -239,66 +246,106 @@ public void fetch() {
+                for (TopicIdPartition tip : acknowledgeRequestState.acknowledgementsMap.keySet()) {
+                    metricsManager.recordFailedAcknowledgements(shareFetchBuffer.getPendingAcknowledgementsCount(tip));
+                    shareFetchBuffer.handleAcknowledgementResponses(tip, Errors.REQUEST_TIMED_OUT);
+                }
+                acknowledgeRequestStates.remove(acknowledgeRequestState);
+            }
+        }
+        PollResult pollResult = null;
+        if (!unsentRequests.isEmpty()) {
+            pollResult = new PollResult(unsentRequests);
+        } else if (acknowledgeRequestStates.isEmpty()) {
+            // Complete the future.
+            if (commitSyncFuture != null) {
+                commitSyncFuture.complete(null);
+            }
+        } else {
+            // Return empty result until all the acknowledgement request states are processed.
+            pollResult = PollResult.EMPTY;
+        }
+        return pollResult;
+    }
+
+    private PollResult processAcknowledgements(long currentTimeMs) {
+        List<UnsentRequest> unsentRequests = new ArrayList<>();
+        for (AcknowledgeRequestState acknowledgeRequestState : acknowledgeRequestStates) {
+            if (acknowledgeRequestState.isExpired()) {
+                acknowledgeRequestStates.remove(acknowledgeRequestState);
+            } else if (!acknowledgeRequestState.retryTimeoutExpired(currentTimeMs)) {
+                if (acknowledgeRequestState.canSendRequest(currentTimeMs)) {
+                    acknowledgeRequestState.onSendAttempt(currentTimeMs);
+                    unsentRequests.add(acknowledgeRequestState.buildRequest(currentTimeMs));
+                }
+            } else {
+                // Fill in TimeoutException
+                for (TopicIdPartition tip : acknowledgeRequestState.acknowledgementsMap.keySet()) {
+                    metricsManager.recordFailedAcknowledgements(shareFetchBuffer.getPendingAcknowledgementsCount(tip));
+                    shareFetchBuffer.handleAcknowledgementResponses(tip, Errors.REQUEST_TIMED_OUT);
+                }
+                acknowledgeRequestStates.remove(acknowledgeRequestState);
+            }
+        }
+        PollResult pollResult = null;
+        if (!unsentRequests.isEmpty()) {
+            pollResult = new PollResult(unsentRequests);
+        } else if (acknowledgeRequestStates.isEmpty()) {
+            // Complete the future.
+            if (commitSyncFuture != null) {
+                commitSyncFuture.complete(null);
+            }
+        } else {
+            // Return empty result until all the acknowledgement request states are processed.
+            pollResult = PollResult.EMPTY;
+        }
+        return pollResult;
+    }
+
     public CompletableFuture<Void> commitSync(final long retryExpirationTimeMs) {
-        final CompletableFuture<Void> result = new CompletableFuture<>();
+        commitSyncFuture = new CompletableFuture<>();
```

Comment:
Probably ought to have some defensive code to ensure that we don't inadvertently have two commitSync operations overlapping. Should never happen, but here you'd overwrite the future and orphan the old one.